crates/cdk-ffi/src/database.rs

```diff
@@ -492,121 +449,956 @@ impl CdkWalletDatabase for WalletDatabaseBridge {
             .into_iter()
-            .map(|tx| tx.try_into())
-            .collect::<Result<Vec<_>, FfiError>>()
-            .map_err(|e| cdk::cdk_database::Error::Database(e.to_string().into()))
+            .map(|info| {
+                Ok(cdk::types::ProofInfo {
+                    proof: info.proof.try_into().map_err(|e: FfiError| {
+                        cdk::cdk_database::Error::Database(e.to_string().into())
+                    })?,
+                    y: info.y.try_into().map_err(|e: FfiError| {
+                        cdk::cdk_database::Error::Database(e.to_string().into())
+                    })?,
+                    mint_url: info.mint_url.try_into().map_err(|e: FfiError| {
+                        cdk::cdk_database::Error::Database(e.to_string().into())
+                    })?,
+                    state: info.state.into(),
+                    spending_condition: info
+                        .spending_condition
+                        .map(|sc| sc.try_into())
+                        .transpose()
+                        .map_err(|e: FfiError| {
+                            cdk::cdk_database::Error::Database(e.to_string().into())
+                        })?,
+                    unit: info.unit.into(),
+                })
+            })
+            .collect();
+
+        cdk_result
     }
+}
 
-    async fn remove_transaction(
+#[async_trait::async_trait]
+impl DbTransactionFinalizer for WalletDatabaseTransactionBridge {
+    type Err = cdk::cdk_database::Error;
+
+    async fn commit(mut self: Box<Self>) -> Result<(), cdk::cdk_database::Error> {
+        self.is_finalized = true;
+        let tx = self.ffi_tx.clone();
+        tx.commit()
+            .await
+            .map_err(|e: FfiError| cdk::cdk_database::Error::Database(e.to_string().into()))
+    }
+
+    async fn rollback(mut self: Box<Self>) -> Result<(), cdk::cdk_database::Error> {
+        self.is_finalized = true;
+        let tx = self.ffi_tx.clone();
+        tx.rollback()
+            .await
+            .map_err(|e: FfiError| cdk::cdk_database::Error::Database(e.to_string().into()))
+    }
+}
+
+impl Drop for WalletDatabaseTransactionBridge {
+    fn drop(&mut self) {
+        if !self.is_finalized {
+            let tx = self.ffi_tx.clone();
+            tokio::spawn(async move {
+                let _ = tx.rollback().await;
+            });
+        }
+    }
+}
+
+pub(crate) struct FfiWalletSQLDatabase<RM>
+where
+    RM: DatabasePool + 'static,
+{
+    inner: SQLWalletDatabase<RM>,
+}
+
+impl<RM> FfiWalletSQLDatabase<RM>
+where
+    RM: DatabasePool + 'static,
+{
+    /// Creates a new instance
+    pub fn new(inner: SQLWalletDatabase<RM>) -> Arc<Self> {
+        Arc::new(Self { inner })
+    }
+}
+
+/// Transaction wrapper for FFI
+pub(crate) struct FfiWalletTransaction {
+    tx: Arc<Mutex<Option<DynWalletDatabaseTransaction>>>,
+}
+
+impl Drop for FfiWalletTransaction {
+    fn drop(&mut self) {
+        let tx = self.tx.clone();
```

Comment:
```suggestion
    fn drop(&mut self) {
        if let Ok(guard) = self.tx.try_lock() {
            if guard.is_none() {
                return;
            }
        }
        let tx = self.tx.clone();
```

We can probally avoid the spawn in most cases.

Reply:
> I think this looks good. But will have to test how it actually works in bindings, specifically with the async spawn in the drop. Have you looked into how this would be handled?

Drop is something that happens in Rust, I guess we should have a runtime when that happens. But I'll test shortly.

Reply:
You were right, I'm debugging now

```
thread '<unnamed>' (91996174) panicked at /Users/cesar/projects/rust/cdk/crates/cdk-common/src/task.rs:14:5:
there is no reactor running, must be called from the context of a Tokio 1.x runtime
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

thread '<unnamed>' (91996174) panicked at /Users/cesar/projects/rust/cdk/crates/cdk-sql-common/src/database.rs:106:13:
there is no reactor running, must be called from the context of a Tokio 1.x runtime
```